src/controllers/auth/protect.ts

```diff
@@ -0,0 +1,103 @@
+import { ENVIRONMENT } from '@/common/config';
+import { JWTExpiresIn } from '@/common/constants';
+import type { CustomRequest } from '@/common/interfaces';
+import { IUser } from '@/common/interfaces';
+import { getFromCache, setCache, setCookie } from '@/common/utils';
+import AppError from '@/common/utils/appError';
+import { catchAsync } from '@/middlewares';
+import { UserModel as User } from '@/models';
+import type { NextFunction, Request, Response } from 'express';
+import jwt from 'jsonwebtoken';
+import { DateTime } from 'luxon';
+import { promisify } from 'util';
+
+export const protect = catchAsync(async (req: Request, res: Response, next: NextFunction) => {
+	// get the cookies from the request headers
+	const { abegAccessToken, abegRefreshToken } = req.cookies as {
+		abegAccessToken: string;
+		abegRefreshToken: string;
+	};
+
+	if (!abegAccessToken || !abegRefreshToken) {
+		throw new AppError('Auth tokens not provided', 401);
+	}
+
+	// verify user access
+
+	const handleUserVerification = async (decoded) => {
+		// fetch user from redis cache or db
+		const cachedUser = (await getFromCache(decoded.id)) as IUser;
+		const user = cachedUser
+			? cachedUser
+			: ((await User.findOne({ _id: decoded.id })
+					.select('refreshToken loginRetries isSuspended isEmailVerified lastLogin')
+					.lean()) as IUser);
+
+		if (!cachedUser && user) {
+			await setCache(decoded.id, user);
+		}
+
+		// check if refresh token matches the stored refresh token in db
+		// in case the user has logged out and the token is still valid
+		// or the user has re authenticated and the token is still valid etc
+
+		if (user.refreshToken !== abegRefreshToken) {
+			throw new AppError('Invalid token. Please log in again!', 401);
+		}
+
+		if (user.isSuspended) {
+			throw new AppError('Your account is currently suspended', 401);
```

Comment:
i think this should explain why i chose a 403

401 Unauthorized:

The 401 status code indicates that the request has not been applied because it lacks valid authentication credentials for the target resource.
The client must authenticate itself to get permission.
In the context of authentication, a 401 response typically means that the user needs to provide valid credentials (e.g., username and password) to access the requested resource.
It is often used when the server requires a user to log in, and the provided credentials are either missing, incorrect, or insufficient.
403 Forbidden:

The 403 status code indicates that the server understood the request but refuses to authorize it.
Unlike 401, the client's identity is known to the server, but the server refuses to fulfill the request.
This status is similar to authorization errors where the client does not have the necessary permissions to access the requested resource.
It is used when authentication is successful, but the authenticated user does not have the necessary permissions (authorization) to perform the requested operation.